ITensorGaussianMPS/src/gmps.jl

```diff
@@ -188,6 +294,193 @@ function givens_rotations(v::AbstractVector{ElT}) where {ElT}
+  of complex numbers in the complex plane, and Givens/Boguliobov Rotations
+  with real arguments only, acting on the interlaced single-particle space of
+  annihilation and creation operator coefficients.
+  """
+function givens_rotations(_v0::Pairing;)
+  v0 = _v0.data
+  N = div(length(v0), 2)
+  if N == 1
+    error(
+      "Givens rotation on 2-element vector not allowed for Pairing-type calculations. This should have been caught elsewhere.",
+    )
+  end
+  ElT = eltype(v0)
+  gs = Circuit{ElT}([])
+  v = copy(v0)
+  r = v[2]
+  ##Given's rotations from creation-operator coefficients
+  gscc, _ = givens_rotations(v[2:2:end])
+  gscc = scale!(gscc, 2)
+  gsca = Circuit(copy(gscc.rotations))
+  gsca = shift!(gsca, -1)
+  gsca = conj!(gsca)
+  gsc = interleave(gscc, gsca)
+  v = gsc * v
+  LinearAlgebra.lmul!(gsc, gs)
+
+  ##Given's rotations from annihilation-operator coefficients
+  gsaa, _ = givens_rotations(v[3:2:end])
+  gsaa = scale!(gsaa, 2)
+  gsaa = shift!(gsaa, +1)
+  gsac = Circuit(copy(gsaa.rotations))
+  gsac = shift!(gsac, +1)
+  gsac = conj!(gsac)
+  gsa = interleave(gsac, gsaa)
+  v = gsa * v
+  LinearAlgebra.lmul!(gsa, gs)
+
+  ##Boguliobov rotation for remaining Bell pair
+  g1, r = givens(v, 2, 3)
+  g2 = Givens(1, 4, g1.c, g1.s')
+  v = g1 * v
+  v = g2 * v #should have no effect
+  LinearAlgebra.lmul!(g2, gs)
+  LinearAlgebra.lmul!(g1, gs)
+  return gs, r
+end
+
+function check_pairing_correlations(Λ0::AbstractMatrix{ElT}) where {ElT<:Number}
+  paired = false
+  Λblocked = reverse_interleave(Λ0)
+  N = div(size(Λblocked, 1), 2)
+  if all(abs.(Λblocked[1:N, (N + 1):end]) .<= eps(Float64))
+    return paired, Λblocked[(N + 1):end, (N + 1):end]
+  else
+    paired = true
+    return paired, Λ0
+  end
+end
+
+function check_pairing_correlations(Λ0::ConservingNf)
+  return Λ0
+end
+
+function check_pairing_correlations(Λ0::Pairing)
+  is_paired, Λ = check_pairing_correlations(Λ0.data)
+  if !is_paired
+    return ConservingNf(Λ)
+  else
+    return Pairing(Λ)
+  end
+end
+
+function get_subblock(
+  _Λ::Pairing,
+  startind::Int;
+  eigval_cutoff::Float64=1e-8,
+  minblocksize::Int=1,
+  maxblocksize::Int=div(size(_Λ.data, 1), 2),
+)
+  blocksize = 0
+  n = 0.0
+  n1 = 0.0
+  n2 = 0.0
+  err = 0.0
+  p = Int[]
+  nB = 0.0
+  uB = 0.0
+  ΛB = 0.0
+  i = startind
+  Λ = _Λ.data
+  N = size(Λ, 1)
+  for blocksize in minblocksize:maxblocksize
+    j = min(2 * i + 2 * blocksize, N)
+    ΛB = @view Λ[(2 * i - 1):j, (2 * i - 1):j]
+    nB, uB = eigen(Hermitian(ΛB))
+    p = sortperm(nB)
+    n1 = nB[first(p)]
+    n2 = nB[last(p)]
+    err = min(abs(n1), abs(n2))
+    n = n1
+    err ≤ eigval_cutoff && break
+  end
+  v = @view uB[:, p[1]]
+  return Pairing(v), Pairing(nB), err
+end
+
+function get_subblock(
+  _Λ::ConservingNf,
+  startind::Int;
+  eigval_cutoff::Float64=1e-8,
+  minblocksize::Int=1,
+  maxblocksize::Int=size(_Λ.data, 1),
+)
+  blocksize = 0
+  n = 0.0
+  err = 0.0
+  p = Int[]
+  nB = 0.0
```

Comment:
This confused me since it's actually of type `Vector{eltype(_Λ)}`. To initialize I think you can just type:
```julia
nB::Vector{eltype(_Λ)}
```

Reply:
Never mind, you would have to use:
```julia
local nB::Vector{eltype(_Λ)}
```
to define a variable without assigning a value (https://docs.julialang.org/en/v1/manual/variables-and-scoping/#Local-Scope). I think assignment is a bit simpler and more canonical so:
```julia
nB = eltype(_Λ)[]
```
seems best.